### ISSUE TITLE
Use `⨟` instead of `|>` (was: Use `∘` instead of `|>`?)

### ISSUE THREAD

**Issue by tkf** [JuliaFolds/Transducers.jl#67]
Ever since I learnt that transducers are adjoint of iterator transforms #13, I've been wanting to rollback the decision to use `|>` instead of `∘`.  We'd then have this nice property (e.g., like we have with matrices _(ABC)' = C'B'A'_):

```julia
(Map(f) ∘ Filter(p) ∘ Cat())'
    = Cat()' ∘ Filter(p)' ∘ Map(f)'
    = flatten ∘ filter(p, _) ∘ Generator(f, _)
```

Other reasons to avoid `|>` are

* We might have an ASCII alias for `∘`: https://github.com/JuliaLang/julia/pull/33573 (one of the motivation behind `|>` was to provide ASCII-based API)
* There was (is?) a plan to deprecate overloading of `|>`: https://github.com/JuliaLang/julia/pull/33684#issuecomment-546708660 / https://github.com/JuliaLang/julia/pull/24886#issuecomment-351825712

This does not mean that the original reason for using `|>` is solved; i.e., the usual order of evaluation due to `∘` (left-to-right) is different from how elements in a collection are processed (right-to-left).  But hiding this from users may just introduce more confusion.

To solve this "flipped `∘`" problem, a better solution may be to use iterator transforms as surface API by

* creating a "shim" iterator transform library which just wraps transducers
* resurrecting https://github.com/JuliaLang/julia/pull/33027 (maybe very hard)
* writing Transducers.jl "backend" for Query.jl (?)

**Comment by jw3126** [comment]
I think there is a way to resolve this problem, by chaning slightly how we think about the category of transducers and iterator transforms.

* There is a single category of `Transforms`. It has a right action on `ReductionFunctoins` and a left action on `Iterators`
* To emphasise the right action on can call it `Transducers`, to emphasise the left action one can call it `IteratorTransforms`. But really there is only a single such category, with two actions.
* This way of thinking leads to very nice notation. We can write
`fold(rf, Filter(pred) ∘ Map(f) ∘ Take(10), itr)`
and it is mathematically unambiguous. 

To move closer to this picture in julia, we should:
* Deprecate `transducer(rf)` and instead introduce `act(rf, transducer)`. With a nice unicode infix symbol if we find one.
* Deprecate `|>`
* Switch the order of `transducer1 ∘ transducer2`

**Comment by jw3126** [comment]
`↫` and `↬` are unicode symbols, that people sometimes use for right/left actions.

**Comment by tkf** [comment]
Thanks a lot for the comments! Very interesting!

> * We can write
>   `fold(rf, Filter(pred) ∘ Map(f) ∘ Take(10), itr)`
>   and it is mathematically unambiguous.

So you mean

```julia
fold(rf, Filter(pred) ∘ Map(f) ∘ Take(10), itr)
== fold(((rf ↫ Filter(pred)) ↫ Map(f)) ↫ Take(10), itr)
== fold(rf, Filter(pred) ↬ Map(f) ↬ Take(10) ↬ itr)
== fold(rf, Filter(pred)(Map(f)(Take(10)(itr))))
== fold(rf, itr |> Take(10) |> Map(f) |> Filter(pred))  # `|>` in the usual sense
```

?  This is beautiful!

In addition to this, I think it'd still be nice to have the "opposite" composition operator where the elements flow left to right. Do you know any symbols used for this purpose?  [Category (mathematics) - Wikipedia](https://en.wikipedia.org/wiki/Category_(mathematics)) mentions "diagrammatic order" composition with `;` (so `g ; f = f ∘ g`) but obviously we can't use `;` as an operator in Julia.  Maybe use `:` as it looks close enough?

```julia
Take(10) : Map(f) : Filter(pred)
=== Filter(pred) ∘ Map(f) ∘ Take(10)
```

or maybe `|`.

It'd be nice if we have other pipe-like ASCII operators in Julia.  I wonder if it is feasible to add `\>` and `<\` to the parser.  We then can have

```julia
fold(rf, itr |> Take(10) |> Map(f) |> Filter(pred))
= fold(rf, itr |> (Take(10) \> Map(f) \> Filter(pred)))
= fold(rf, Take(10) \> Map(f) \> Filter(pred), itr)
= fold(rf ↫ Take(10) \> Map(f) \> Filter(pred), itr)
= fold(rf, (Filter(pred) <\ Map(f) <\ Take(10)) <| itr)
= fold(rf, (Filter(pred) <\ Map(f) <\ Take(10)) ↬ itr)
```

**Comment by jw3126** [comment]
> So you mean
> 
> ```julia
> fold(rf, Filter(pred) ∘ Map(f) ∘ Take(10), itr)
> == fold(((rf ↫ Filter(pred)) ↫ Map(f)) ↫ Take(10), itr)
> == fold(rf, Filter(pred) ↬ Map(f) ↬ Take(10) ↬ itr)
> == fold(rf, Filter(pred)(Map(f)(Take(10)(itr))))
> == fold(rf, itr |> Take(10) |> Map(f) |> Filter(pred))  # `|>` in the usual sense
> ```
> 
Jup, this is what I meant.
> 
> In addition to this, I think it'd still be nice to have the "opposite" composition operator where the elements flow left to right. Do you know any symbols used for this purpose? [Category (mathematics) - Wikipedia](https://en.wikipedia.org/wiki/Category_(mathematics)) mentions "diagrammatic order" composition with `;` (so `g ; f = f ∘ g`) but obviously we can't use `;` as an operator in Julia. Maybe use `:` as it looks close enough?

I also often wished that julia, had an operator for opposite composition. The only math notation I am aware of is `∘ₒₚ` . But it looks a bit ugly in unicode. Maybe there are other programming languages with opposite composition operator? 
*  `\>` and `<\` they are a cool pair of operators. But I don't really like them for composition. I think `∘` is so widely used, it should be one of the composition operators, it just needs a partner. (I would not like to have an infix alias for `∘`)
*  `:` it is already used for ranges, so using it for this totally different purpose is undesirable
* `|` is taken for `or` and I associate it more with `|>` then with `∘`

**Comment by tkf** [comment]
> Maybe there are other programming languages with opposite composition operator?

A quick googling told me that Haskell's [Control.Arrow](https://hackage.haskell.org/package/base-4.12.0.0/docs/Control-Arrow.html) has `>>>` (= `∘ₒₚ`) and `<<<` (= `∘`).  I think `\>` and `<\` are similar in that they clarifies the directions.

> I think `∘` is so widely used, it should be one of the composition operators

I agree with this part. I wasn't suggesting to drop `∘` (although I thought `<\` was a nice ASCII alias of `∘`).

To stick with something like `∘`, maybe something like `~∘` or `-∘` makes sense? Or even `'∘` or `∘'` although parsing them sounds hard.

**Comment by tkf** [comment]
> obviously we can't use `;` as an operator in Julia

I was wrong :laughing: 

```julia
julia> Base.vcat(xfs::Transducer...) = foldl(Base.:|>, xfs)

julia> [Map(sin); Map(cos); Map(tan)]
Map(sin) |>
    Map(cos) |>
    Map(tan)
```

(No I'm not serious)

**Comment by tkf** [comment]
> if it is feasible to add `\>` and `<\` to the parser

It was super easy https://github.com/tkf/julia/commit/56d2556a57e311416f2650dcbf03fb469a0df9ec

**Comment by jw3126** [comment]
I just realized, that one can do right now in julia:
```julia
julia> ∘ₒₚ(g,f) = f∘g
∘ₒₚ (generic function with 1 method)
```
Maybe use `∘, ∘ₒₚ` with `<\, \>` as ASCII aliases?

**Comment by tkf** [comment]
That sounds good.

I opened a PR to request adding the syntax https://github.com/JuliaLang/julia/pull/34340. I'm not requesting adding the functions yet as I thought it may slow down the reviewing.

**Comment by tkf** [comment]
I'm thinking to create a super simple package (say Compositions.jl) with

```julia
const compose = ∘
flip(f) = (args...) -> f(reverse(args)...)
```

If we have JuliaLang/julia#34340 but `∘ₒₚ` etc. are rejected then it can also contain

```julia
∘ₒₚ(args...) = ∘(reverse(args)...)
const var"</" = ∘
const var"/>" = ∘ₒₚ
```

I think it's nice to have ASCII functions `compose` and `flip(compose)` even in old `julia`s.  What do you think?

Also, is `filp` better be called `op`?  Though then I guess we'd want `op` to be a functor.  But that's impossible unless we can write `op(::Morphism)` and `op(::Object)` or something like that.

**Comment by jw3126** [comment]
`Compositions.jl` sounds nice. Yeah, from `op` I would expect that it is a function. (Which would require that `Base.∘` returns a callable struct and not a lambda. Which would also enable better e.g `show(isfinite ∘ maximum)`.

**Comment by tkf** [comment]
BTW I've been wanting to define `show` for `f ∘ g` for a while now too.

**Comment by jw3126** [comment]
https://discourse.julialang.org/t/add-composition-operator-unicode-2a1f-for-flipped/34436

**Comment by tkf** [comment]
So, I've been wondering what the API of the "right action" of transducers (action on reducing functions) should be.  It's a bit unfortunate that the left action is so easy (`xf(itr)`) while the right action is a bit clumsy (`reducingfunction(xf, rf)`).  I think the cleanest solution might be to use something like

```julia
adjoint(xf::Transducer) = rf -> reducingfunction(xf, rf)
```

This way, both `Map(f)(itr)` and `Map(f)'(rf)` work nicely and cleanly.

(`adjoint` probably should return `ReducingFunctionTransform` object or something so that `xf''` is `xf` and the composition works in this category too)

**Comment by tkf** [comment]
This is now implemented by #319.  The new documentation (preview) is here: https://juliafolds.github.io/Transducers.jl/previews/PR319/

In particular, it's somewhat unfortunate (though expected) now that the notion of "transducer" is diverged from the rest of world: https://juliafolds.github.io/Transducers.jl/previews/PR319/explanation/glossary/#Transducer-1 (glossary).  But I think practicality of `itr |> xf1 |> xf2` is a win overall.

@jw3126 Want to have a look at it?

**Comment by jw3126** [comment]
Cool thanks, I will take a look this week.
